src/refract/Element.h

```diff
@@ -2,798 +2,186 @@
-        {
-        }
-
-        RefElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-    };
-
-    struct ExtendElementTrait : public ElementCollectionTrait<> {
-        static const std::string element()
-        {
-            return "extend";
-        }
-    };
-
-    struct ExtendElement : Element<ExtendElement, ExtendElementTrait> {
-        ExtendElement() : Type()
-        {
-        }
-
-        ExtendElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-
-        void push_back(IElement* e);
-        void set(const ValueType& val);
-
-        IElement* merge() const;
-    };
-
-    struct OptionElementTrait : public ElementCollectionTrait<> {
-        static const std::string element()
-        {
-            return "option";
-        }
-    };
-
-    struct OptionElement : Element<OptionElement, OptionElementTrait> {
-
-        OptionElement() : Type()
-        {
-        }
-
-        OptionElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
-
-        OptionElement(IElement* e) : Type()
-        {
-            value.push_back(e);
-        }
-
-        void push_back(IElement* e)
-        {
-            // FIXME:
-            // probably add check for allowed type
-            hasContent = true;
-            value.push_back(e);
-        }
-    };
-
-    struct SelectElementTrait : public ElementCollectionTrait<OptionElement> {
-        static const std::string element()
-        {
-            return "select";
-        }
-
-        static void cloneValue(const ValueType& self, ValueType& other)
-        {
-            for (std::vector<OptionElement*>::const_iterator it = self.begin(); it != self.end(); ++it) {
-                IElement* clonned = (*it)->clone();
-                other.push_back(static_cast<OptionElement*>(clonned));
-            }
-        }
-    };
-
-    struct SelectElement : Element<SelectElement, SelectElementTrait> {
-        SelectElement() : Type()
-        {
-        }
+    ///
+    /// Create an empty Element of given type
+    /// @remark an empty Element has an empty data structure definition (DSD)
+    ///
+    template <typename ElementT>
+    auto make_empty()
+    {
+        return std::make_unique<ElementT>();
+    }
+
+    ///
+    /// Create an Element of given type forwarding arguments to the constructor
+    /// of its data structure definition (DSD)
+    ///
+    template <typename ElementT, typename... Args>
+    auto make_element(Args&&... args)
+    {
+        using DataT = typename ElementT::ValueType;
+        return std::make_unique<ElementT>(DataT{ std::forward<Args>(args)... });
+    }
 
-        SelectElement(const TraitType::ValueType& value) : Type()
-        {
-            set(value);
-        }
+    template <typename Primitive, typename DataT = typename data::data_of<Primitive>::type>
+    auto make_primitive(const Primitive& p)
```

Comment:
In earlier comment i suggest to rename `from_primitive` 
Reason is existence of `RefractToPrimitive()` and it is confising "what primitive"
I understand it is supplementary function to `is_primitive` but i thing it is confusing (at least for me)

Reply:
I intended `make_primitive` to be a variant of `make_element` where the type of the element is deduced because it has a C++ primitive equivalent; e. g. `make_primitive(std::string{})`, `make_primitive("hello world")`, `make_primitive(42)` instead of `make_element<StringElement>(std::string{})`, et cetera. This construct already existed and unfortunately is used generically at some places in the code base.
Where do you see `RefractToPrimitive` in this branch?

Reply:
Sry mistake it is named `PrimitiveToRefract`

I understand to argue, but `make_primitive` evoke something diferent than "create element from primitive". At least for me :)

Reply:
I've looked at the code after renaming to `from_primitive`, and it reads alright.